client/app/lib/redux/modules/sidebar/stacks.coffee

```diff
@@ -0,0 +1,306 @@
+EDIT = expandActionType withNamespace 'EDIT'
+INITIALIZE = expandActionType withNamespace 'INITIALIZE'
+{ LOAD, REMOVE } = require 'app/redux/modules/bongo'
+
+bongo = require 'app/redux/modules/bongo'
+
+reducer = (state = immutable({}), action) ->
+
+  switch action.type
+
+    when 'STACK_REVISION_SUCCESS'
+      { result: { _id, data, error } } = action
+      return state  if error
+      state = state.set _id, immutable data
+
+      return state
+
+    else
+      return state
+
+# actions
+
+initializeStack = (template) -> (dispatch) ->
+  { reactor } = kd.singletons
+
+  waterfall([
+    (next) ->
+      dispatch {
+        types: [LOAD.BEGIN, LOAD.SUCCESS, LOAD.FAIL]
+        bongo: -> template.generateStack().then (result) ->
+          { stack, results: { machines } } = result
+          instances = []
+          instances.push stack
+          machines.forEach (machine) -> instances.push machine.obj
+          next null, machines[0].obj
+          reactor.dispatch 'GENERATE_STACK_SUCCESS', { template, stack }
+
+          return instances
+        }
+
+    (machine, next) ->
+      console.log '???? ', machine, ' should be null amk', machine.label
+      new kd.NotificationView { title: 'Stack generated successfully' }
+      reloadIDE machine.label
+      next()
+  ])
+  console.log 'something'
+
+
+openOnGitlab = (stack) ->
+
+  remoteUrl = stack.getAt ['config', 'remoteDetails', 'originalUrl']
+  kd.singletons.linkController.openOrFocus remoteUrl
+
+
+handleRoute = (route) ->
+
+  kd.singletons.router.handleRoute route
+
+
+setAccess = (template) ->
+
+  return  unless template
+
+  return {
+    types: 'UPDATE'
+    promise: -> new Promise (resolve, reject) ->
+      template.setAccess 'group', (err) ->
+        return resolve()  unless err
+        reject()
+  }
+
+makeTeamDefault = (template, credential) ->
+
+  kd.singletons.computeController.makeTeamDefault2 template, credential, ->
+    console.log '??? make team default'
+
+
+destroyStack = (stack, machines, type = 'deleteStack') ->
+
+  return  unless stack
+
+  { computeController, appManager } = kd.singletons
+
+  return {
+    types: [REMOVE.BEGIN, REMOVE.SUCCESS, REMOVE.FAIL]
+    promise: -> new Promise (resolve, reject) ->
+      computeController.ui.askFor type, {}, (status) ->
+
+        return  unless status.confirmed # reject
+
+        appManager.quitByName 'IDE', ->
+          computeController.destroyStack stack, (err) ->
+
+            return  if showError err # reject
+            result = []
+            result.push stack
+
+            machines.forEach (machine) -> result.push machine
+            handleRoute '/IDE'
+            return resolve result
+
+          , followEvents = no
+  }
+
+
+reloadIDE = (machineSlug) ->
+  console.log 'machine Slug ', machineSlug
+  { computeController } = kd.singletons
+  computeController.reloadIDE machineSlug
+
+
+reinitStack = (stack, machines, template) -> (dispatch) ->
```

Comment:
I don't line changing the global `action` signature. Maybe we should pass the `dispatch` function with `getState` in promise middleware [here].(https://github.com/koding/koding/blob/master/client/app/lib/redux/middleware/promise.coffee#L30)

Something like `promise({ dispatch: store.dispatch, getState: store.getState })`

So that our `promise` keys can get `dispatch` and `getState` injected.

That way what we would do is:
```coffee
reinitStack = (stack, machines, template) ->
  return {
    types: [REINIT.BEGIN, REINIT.SUCCESS, REINIT.FAIL]
    promise: ({ dispatch }) ->
      dispatch(destroyStack stack, machines, 'reinitStack').then ->
        dispatch(initializeStack(template) # this requires a change to `initializeStack` as well.
  }
```